MonoGame.Framework/Graphics/GraphicsDevice.cs

```diff
@@ -509,6 +514,41 @@ public DepthStencilState DepthStencilState
 
                 _depthStencilStateDirty = true;
             }
+        }
+
+        /// <summary>
+        /// Get an effect and store it in the cache if necessary.
+        /// </summary>
+        /// <param name="graphicsDevice"></param>
+        /// <param name="effectCode"></param>
+        /// <param name="index"></param>
+        /// <param name="count"></param>
+        /// <param name="effectKey"></param>
+        /// <param name="headerSize"></param>
+        /// <returns></returns>
+        internal Effect GetEffect(GraphicsDevice graphicsDevice, byte[] effectCode, int index, int count, int effectKey, int headerSize)
+        {
+            Effect cloneSource;
+
+            lock (_effectCacheLock)
+            {
+                //First look for it in the cache.
+                if (!EffectCache.TryGetValue(effectKey, out cloneSource))
+                {
+                    using (var stream = new MemoryStream(effectCode, index + headerSize, count - headerSize, false))
+                    using (var reader = new BinaryReader(stream))
+                    {
+                        // Create one.
+                        cloneSource = new Effect(graphicsDevice);
+                        cloneSource.ReadEffect(reader);
+
+                        // Cache the effect for later in its original unmodified state.
+                        graphicsDevice.EffectCache.Add(effectKey, cloneSource);
```

Comment:
... while here we add the effectKey to graphicsDevice.EffectCache, which can be a different object.

Reply:
That's an excellent point, I'll go change that.